osmoutils/osmocli/parsers.go

```diff
@@ -0,0 +1,282 @@
+	case reflect.Uint16:
+		fallthrough
+	case reflect.Uint32:
+		fallthrough
+	case reflect.Uint64:
+		fallthrough
+	case reflect.Uint:
+		u, err := ParseUint(arg, fType.Name)
+		if err != nil {
+			return true, err
+		}
+		fVal.SetUint(u)
+		return true, nil
+	// SetInt allows anyof type i8,i16,i32,i64 and int
+	case reflect.Int8:
+		fallthrough
+	case reflect.Int16:
+		fallthrough
+	case reflect.Int32:
+		fallthrough
+	case reflect.Int64:
+		fallthrough
+	case reflect.Int:
+		typeStr := fType.Type.String()
+		var i int64
+		var err error
+		if typeStr == "time.Duration" {
+			dur, err2 := time.ParseDuration(arg)
+			i, err = int64(dur), err2
+		} else {
+			i, err = ParseInt(arg, fType.Name)
+		}
+		if err != nil {
+			return true, err
+		}
+		fVal.SetInt(i)
+		return true, nil
+	case reflect.String:
+		s, err := ParseDenom(arg, fType.Name)
+		if err != nil {
+			return true, err
+		}
+		fVal.SetString(s)
+		return true, nil
+	case reflect.Ptr:
+	case reflect.Slice:
+		typeStr := fType.Type.String()
+		if typeStr == "types.Coins" {
+			coins, err := ParseCoins(arg, fType.Name)
+			if err != nil {
+				return true, err
+			}
+			fVal.Set(reflect.ValueOf(coins))
+			return true, nil
+		}
+	case reflect.Struct:
+		typeStr := fType.Type.String()
+		if typeStr == "types.Coin" {
+			coin, err := ParseCoin(arg, fType.Name)
+			if err != nil {
+				return true, err
+			}
+			fVal.Set(reflect.ValueOf(coin))
+			return true, nil
+		}
+	}
+	fmt.Println(fType.Type.Kind().String())
+	return true, fmt.Errorf("field type not recognized. Got type %v", fType)
+}
+
+func ParseUint(arg string, fieldName string) (uint64, error) {
+	v, err := strconv.ParseUint(arg, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("could not parse %s as uint for field %s: %w", arg, fieldName, err)
+	}
+	return v, nil
+}
+
+func ParseInt(arg string, fieldName string) (int64, error) {
+	v, err := strconv.ParseInt(arg, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("could not parse %s as int for field %s: %w", arg, fieldName, err)
+	}
+	return v, nil
+}
+
+func ParseUnixTime(arg string, fieldName string) (time.Time, error) {
+	timeUnix, err := strconv.ParseInt(arg, 10, 64)
+	if err != nil {
+		return time.Time{}, fmt.Errorf("could not parse %s as unix time for field %s: %w", arg, fieldName, err)
+	}
+	startTime := time.Unix(timeUnix, 0)
+	return startTime, nil
+}
+
+func ParseDenom(arg string, fieldName string) (string, error) {
+	return strings.TrimSpace(arg), nil
+}
+
+// TODO: Make this able to read from some local alias file for denoms.
+func ParseCoin(arg string, fieldName string) (sdk.Coin, error) {
+	coin, err := sdk.ParseCoinNormalized(arg)
+	if err != nil {
+		return sdk.Coin{}, fmt.Errorf("could not parse %s as sdk.Coin for field %s: %w", arg, fieldName, err)
+	}
+	return coin, nil
+}
+
+// TODO: Make this able to read from some local alias file for denoms.
+func ParseCoins(arg string, fieldName string) (sdk.Coins, error) {
+	coins, err := sdk.ParseCoinsNormalized(arg)
```

Comment:
There are some cases where `sdk.ParseCoinsNormalized` doesn't work. For example, here:
https://github.com/osmosis-labs/osmosis/blob/da6b7d1ab3cc8a71294e81896dd24c2b0504ac1d/x/gamm/client/cli/tx.go#L487

Is that case handled in this PR?

Reply:
yeah, explicitly didn't replace those for now. Need to add a way to set a custom per-field parser to get those replaced